UnitTests/GitCommands.Tests/Settings/AppSettingsTests.cs

```diff
@@ -0,0 +1,641 @@
+            Assert.That(storedValue, Is.EqualTo(defaultValue));
+        }
+
+        [Test]
+        [TestCaseSource(nameof(TestCases))]
+        public void Should_save_value(string propertyName, object value, object defaultValue, bool isSetting)
+        {
+            // Arrange
+            object root = null;
+            var property = typeof(AppSettings).GetProperty(propertyName);
+
+            if (isSetting)
+            {
+                root = typeof(AppSettings).GetProperty(propertyName)
+                    .GetValue(null);
+
+                property = property.PropertyType
+                    .GetProperty(nameof(ISetting<string>.Value));
+            }
+
+            var filePath = Path.GetTempFileName();
+
+            File.WriteAllText(filePath, SettingsFileContent);
+
+            var container = new RepoDistSettings(null, GitExtSettingsCache.Create(filePath), SettingLevel.Unknown);
+            object storedValue = null;
+
+            // Act
+            AppSettings.UsingContainer(container, () =>
+            {
+                property.SetValue(root, value);
+
+                storedValue = property.GetValue(root);
+            });
+
+            // Assert
+            if (Type.GetTypeCode(property.PropertyType) == TypeCode.String)
+            {
+                if (isSetting)
+                {
+                    Assert.That(storedValue, Is.EqualTo(value ?? string.Empty));
+                }
+                else
+                {
+                    Assert.That(storedValue, Is.EqualTo(value ?? defaultValue));
+                }
+            }
+            else
+            {
+                Assert.That(storedValue, Is.EqualTo(value));
+            }
+        }
+
+        #region Test Cases
+
+        private static IEnumerable<object[]> TestCases()
+        {
+            foreach (var value in Values())
+            {
+                foreach ((string propertyName, object defaultTypeValue, object defaultValue, bool isSetting) in PropertyDescriptions())
+                {
+                    var property = typeof(AppSettings).GetProperty(propertyName);
+
+                    if (value is null)
+                    {
+                        if (defaultTypeValue is null)
+                        {
+                            yield return new object[] { propertyName, value, defaultValue, isSetting };
+                        }
+                    }
+                    else
+                    {
+                        var valueType = Nullable.GetUnderlyingType(value.GetType()) ?? value.GetType();
+                        var propertyType = Nullable.GetUnderlyingType(property.PropertyType) ?? property.PropertyType;
+
+                        if (valueType == propertyType)
+                        {
+                            yield return new object[] { propertyName, value, defaultValue, isSetting };
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
💭  I don't typically like using local functions, but I think here is a good scenario to use them and move all the utility methods into `TestCases` so they don't pollute the class namespace.